include/libdnf/conf/option_enum.hpp

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright (C) 2018-2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_CONF_OPTION_ENUM_HPP
+#define LIBDNF_CONF_OPTION_ENUM_HPP
+
+#include "option.hpp"
+
+#include <functional>
+#include <vector>
+
+namespace libdnf {
+
+template <typename T>
+class OptionEnum : public Option {
+public:
+    using ValueType = T;
+    using FromStringFunc = std::function<ValueType(const std::string &)>;
+
+    class InvalidValue : public Option::InvalidValue {
+    public:
+        using Option::InvalidValue::InvalidValue;
+        const char * get_domain_name() const noexcept override { return "libdnf::OptionEnum"; }
+    };
+
+    class NotAllowedValue : public InvalidValue {
+    public:
+        using InvalidValue::InvalidValue;
+        const char * get_name() const noexcept override { return "NotAllowedValue"; }
+        const char * get_description() const noexcept override { return "Not allowed value"; }
+    };
+
+    OptionEnum(ValueType default_value, const std::vector<ValueType> & enum_vals);
+    OptionEnum(ValueType default_value, std::vector<ValueType> && enum_vals);
+    OptionEnum(ValueType default_value, const std::vector<ValueType> & enum_vals, FromStringFunc && from_string_func);
+    OptionEnum(ValueType default_value, std::vector<ValueType> && enum_vals, FromStringFunc && from_string_func);
+    OptionEnum * clone() const override;
+    void test(ValueType value) const;
+    ValueType from_string(const std::string & value) const;
+    void set(Priority priority, ValueType value);
+    void set(Priority priority, const std::string & value) override;
+    T get_value() const;
+    T get_default_value() const;
+    std::string to_string(ValueType value) const;
```

Comment:
Functions test, from_string, and to_string looks like internal function, like protected. Do you know any external usage?

Reply:
I think I had usage in old libdnf. But I'm not sure now. I will look at it. I can move the functions to the protected section. And we will can make they public in future if it is needed.